### PR TITLE
Openstorage changes for new fsck CLI - List snaps, Delete snaps, List vols

### PR DESCRIPTION
**Pull request by ppadalia**
**What this PR does / why we need it**:  
Openstorage changes for new fsck CLI - List snaps, Delete snaps, List vols

**Which issue(s) this PR fixes** (optional)  
PWX-33965

**Testing Notes**  
N/A - Will be uploaded on porx PR
https://github.com/portworx/porx/pull/12350

## list-snapshots ##
```
14:04:13  time=2023-10-20T08:34:13Z level=info msg=SSHCMD: [root@10.13.187.146:22]# sudo sudo /opt/pwx/bin/pxctl volume check list-snapshots volumeFsckListDelSnaps1
14:04:13  time=2023-10-20T08:34:13Z level=info msg=SSHOUT: [root@10.13.187.146:22]#
14:04:13  Volume Snapshot ID	Volume Snapshot Name
14:04:13  764638784223047136	volumeFsckListDelSnaps1_fsck_snap_2023_Oct_20_08_34_01
14:04:13  963046767004546850	volumeFsckListDelSnaps1_fsck_snap_2023_Oct_20_08_33_49

14:04:13  time=2023-10-20T08:34:13Z level=info msg=SSHCMD: [root@10.13.187.146:22]# sudo sudo /opt/pwx/bin/pxctl volume check list-snapshots volumeFsckListDelSnaps1 -j
14:04:13  time=2023-10-20T08:34:13Z level=info msg=SSHOUT: [root@10.13.187.146:22]#
14:04:13  {
14:04:13   "snapshots": {
14:04:13    "764638784223047136": {
14:04:13     "volume_snapshot_name": "volumeFsckListDelSnaps1_fsck_snap_2023_Oct_20_08_34_01"
14:04:13    },
14:04:13    "963046767004546850": {
14:04:13     "volume_snapshot_name": "volumeFsckListDelSnaps1_fsck_snap_2023_Oct_20_08_33_49"
14:04:13    }
14:04:13   }
14:04:13  }
```

## delete-snapshots ##
```
14:04:13  time=2023-10-20T08:34:13Z level=info msg=SSHCMD: [root@10.13.187.146:22]# sudo sudo /opt/pwx/bin/pxctl volume check delete-snapshots volumeFsckListDelSnaps1
14:04:14  time=2023-10-20T08:34:13Z level=info msg=SSHOUT: [root@10.13.187.146:22]#
14:04:14  Filesystem-check delete-snapshots request completed. All fsck created snapshots for the specified volume will be deleted.
14:04:14  Issue 'pxctl volume check list-snapshots <volname>' or check cluster alerts to confirm the deletion of snapshots.
14:04:14  
```

## list-volumes ##
```
12:16:44  time=2023-10-20T06:46:43Z level=info msg=SSHCMD: [root@10.13.181.170:22]# sudo sudo /opt/pwx/bin/pxctl volume check list-volumes
12:16:44  time=2023-10-20T06:46:43Z level=info msg=SSHOUT: [root@10.13.181.170:22]#
12:16:44  Volume ID		Volume Name		Health Status				Volume Has FS Errors
12:16:44  292449395648734571	volumeFsckListVolumes_7	FS_HEALTH_STATUS_SAFE_TO_FIX		no
12:16:44  175374481142876324	volumeFsckListVolumes_5	FS_HEALTH_STATUS_SAFE_TO_FIX		no
12:16:44  955821806768222836	volumeFsckListVolumes_1	FS_HEALTH_STATUS_NEEDS_INSPECTION	no
12:16:44  504247086299902947	volumeFsckListVolumes_9	FS_HEALTH_STATUS_SAFE_TO_FIX		no
12:16:44  204242062023435478	volumeFsckListVolumes_3	FS_HEALTH_STATUS_NEEDS_INSPECTION	no

time=2023-10-20T06:46:43Z level=info msg=SSHCMD: [root@10.13.181.170:22]# sudo sudo /opt/pwx/bin/pxctl volume check list-volumes -j
12:16:44  time=2023-10-20T06:46:43Z level=info msg=SSHOUT: [root@10.13.181.170:22]#
12:16:44  {
12:16:44   "volumes": {
12:16:44    "175374481142876324": {
12:16:44     "volume_name": "volumeFsckListVolumes_5",
12:16:44     "health_status": 2
12:16:44    },
12:16:44    "204242062023435478": {
12:16:44     "volume_name": "volumeFsckListVolumes_3",
12:16:44     "health_status": 3
12:16:44    },
12:16:44    "292449395648734571": {
12:16:44     "volume_name": "volumeFsckListVolumes_7",
12:16:44     "health_status": 2
12:16:44    },
12:16:44    "504247086299902947": {
12:16:44     "volume_name": "volumeFsckListVolumes_9",
12:16:44     "health_status": 2
12:16:44    },
12:16:44    "955821806768222836": {
12:16:44     "volume_name": "volumeFsckListVolumes_1",
12:16:44     "health_status": 3
12:16:44    }
12:16:44   }
12:16:44  }
```

**Special notes for your reviewer**:  
N/A

### CODE REVIEW

File: api/api.proto [hunk 1]
```diff
@@ -1827,6 +1827,15 @@ service OpenStorageFilesystemTrim {
   //    `OpenStorageFilesystemCheckClient.Start(Mode='fix_safe')`
   //    Status of this operation can be polled in the way mentioned in step 3
   //    This operation can be stopped a Stop request as mentioned in step 6
+  // 8. To list all snapshots in the filesystem that are created by fsck, issue a
+  //	grpc call to
+  //	`OpenStorageFilesystemCheckClient.ListSnapshots()`
+  // 9. To delete all snapshots in the filesystem that are created vy fsck, issue a
```
Comment (sulakshm) [inline api/api.proto:1833]: s/vy/by

File: api/api.proto [hunk 2]
```diff
@@ -1856,7 +1865,33 @@ service OpenStorageFilesystemCheck {
...
+        post: "/v1/filesystem-check/delete-snapshots"
+        body: "*"
+      };
+    }
+
+  // List of all volumes which require fsck check/fix to be run
+  rpc ListVolumes(SdkFilesystemCheckListVolumesRequest)
+    returns (SdkFilesystemCheckListVolumesResponse){
+      option(google.api.http) = {
+        post: "/v1/filesystem-check/list-volumes"
```
Comment (sulakshm) [inline api/api.proto:1890]: why is list-volumes a post request and not a get?

File: api/api.proto [hunk 3]
```diff
@@ -5408,6 +5461,44 @@ message SdkFilesystemCheckStopRequest {
...
+message SdkFilesystemCheckListVolumesRequest{
+}
+
+// SdkFilesystemCheckListVolumesResponse defines a response to list
+// all volumes needing fsck check/fix
+message SdkFilesystemCheckListVolumesResponse{
+  // Map of volume ids and volume info
+  map<string, FilesystemCheckVolInfo> volumes = 1;
+}
+
```
Comment (sulakshm) [inline api/api.proto:5501]: a sample log here in the PR, as to what is the request and what is the output like will help.

File: api/server/sdk/filesystem_check.go [hunk 4]
```diff
@@ -93,3 +93,54 @@ func (s *FilesystemCheckServer) Stop(
...
+func (s *FilesystemCheckServer) ListVolumes(
+	ctx context.Context,
+	req *api.SdkFilesystemCheckListVolumesRequest,
+) (*api.SdkFilesystemCheckListVolumesResponse, error) {
+	if s.driver(ctx) == nil {
+		return nil, status.Error(codes.Unavailable, "Resource has not been initialized")
+	}
+	var err error
+
+	r, err := s.driver(ctx).FilesystemCheckListVolumes(req)
```
Comment (sulakshm) [inline api/server/sdk/filesystem_check.go:143]: would this list all the volumes in the cluster that need fsck or only those in the local node?

**Review by sulakshm — COMMENTED** [review]
except for one request submitted as POST vs GET rest all are minor nits from my end.

File: api/api.proto (hunk 1 above)
Comment (ppadalia) [reply api/api.proto]: Done

File: api/api.proto (hunk 2 above)
Comment (ppadalia) [reply api/api.proto]: Fixed

File: api/api.proto (hunk 3 above)
Comment (ppadalia) [reply api/api.proto:5500]: Added sample output in PR.

File: api/server/sdk/filesystem_check.go (hunk 4 above)
Comment (ppadalia) [reply api/server/sdk/filesystem_check.go:143]: All volumes in cluster

**Review by sulakshm — APPROVED** [review]

**Pull request merged**